Prometheus Input Package (#6170)
elastic/integrations: packages/prometheus_input/docs/README.md

        },
        "prometheus_target_interval_length_seconds_count": {
            "counter": 1,
            "rate": 0
        },
        "prometheus_target_interval_length_seconds_sum": {
            "counter": 15.000401344,
            "rate": 0
        },
        "prometheus_tsdb_compaction_chunk_range_seconds": {
            "histogram": {
                "values": [50, 300, 1000, 4000, 16000],
                "counts": [10, 2, 34, 7]
            }
        }
    },
}

```

#### Filtering metrics

In order to filter out/in metrics one can make use of `Metrics Filters Include`, `Metrics Filters Exclude` settings:

```yml
Metrics Filters Include: ["node_filesystem_*"]
Metrics Filters Exclude: ["node_filesystem_device_*"]
```

The configuration above will include only metrics that match `node_filesystem_*` pattern and do not match `node_filesystem_device_*`.

To keep only specific metrics, anchor the start and the end of the regexp of each metric:

- the caret ^ matches the beginning of a text
- the dollar $ matches the end of a text

```yml
Metrics Filters Include: ["^node_network_net_dev_group$", "^node_network_up$"]
```

### Datastream Dataset Name

> The users of the Prometheus Input Package have the option of adding their own dataset name, to which the events get added. Prometheus Metrics from different services can be collected by adding multiple instances of Input package. The metrics can be filtered on the basis of dataset name.

Reviewer: Does this mean that if they dont put something in the field, the field has a default value? 
No is not it?

So I guess the dataset name should go up in the Readme and be a required option for the user to fill in

Author: Yes, there is no default value by default. And this is a must field. 
I think to be on safer side we can have a default value for this as well.

# Prometheus Input Package

This input package can collect metrics from:
- [Prometheus Exporters (Collectors)](#prometheus-exporters-collectors).
It gives users the flexibility to add custom mappings and ingest pipelines.

## Metrics

### Prometheus Exporters (Collectors)

> The Prometheus input package connects to the Prometheus server and pulls metrics using the `/metrics` endpoint.

Reviewer 2: Metricbeat uses a better/correct description:
> To scrape metrics from a Prometheus exporter

The existing integration says though, is it in-accurate then?
>Prometheus integration collector dataset connects to the Prometheus server

We are scrapping from an exporter or Prometheus end point.

> The Prometheus input package connects to the Prometheus server and pulls metrics using the `/metrics` endpoint.

Author: We scrape from Prometheus Exporters. So I would rename this to exporter

> The Prometheus input package connects to the Prometheus server and pulls metrics using the `/metrics` endpoint.

Author: Actually since this is only collector based, I would rather not have this line in the readme, since we are specifying the metrics collected from exporter before this

#### Scraping from a Prometheus exporter

To scrape metrics from a Prometheus exporter, configure the `hosts` setting to it. The path
to retrieve the metrics from (`/metrics` by default) can be configured with Metrics Path.

#### Histograms and types

> `Use Types` parameter (default: true) enables a different layout for metrics storage, leveraging Elasticsearch

Reviewer 3: ```suggestion
`Use Types` parameter (default: `true`) enables a different layout for metrics storage, leveraging Elasticsearch
```

types, including {{ url "elasticsearch-histograms" "histograms" }}.

> `Rate Counters` parameter (default: true) enables calculating a rate out of Prometheus counters. When enabled, integration stores

Reviewer 3: ```suggestion
`Rate Counters` parameter (default: `true`) enables calculating a rate out of Prometheus counters. When enabled, integration stores
```

the counter increment since the last collection. This metric should make some aggregations easier and with better

> performance. This parameter can only be enabled in combination with `Use Types`.

Reviewer 3: > This metric should make some aggregations easier and with better performance

I want to understand this. The addition of this metric make aggregation easier but we get better performance for what? Is it aggregation?

Author: We know about the better aggregation. So maybe, we can have only that for now for input package

When `Use Types` and `Rate Counters` are enabled, metrics are stored like this:

```json
{
    "prometheus": {
        "labels": {
            "instance": "172.27.0.2:9090",
            "job": "prometheus"
        },
        "prometheus_target_interval_length_seconds_count": {
            "counter": 1,
            "rate": 0
        },
        "prometheus_target_interval_length_seconds_sum": {
            "counter": 15.000401344,
            "rate": 0
        },
        "prometheus_tsdb_compaction_chunk_range_seconds": {
            "histogram": {
                "values": [50, 300, 1000, 4000, 16000],
                "counts": [10, 2, 34, 7]
            }
        }
    },
}

```

> #### Scraping all metrics from a Prometheus server

Author: Actually I need to remove this line as we are not supporting remote_write in input package.

We recommend using the Remote Write dataset for this, and make Prometheus push metrics to Agent.

#### Filtering metrics

In order to filter out/in metrics one can make use of `Metrics Filters Include`, `Metrics Filters Exclude` settings:

```yml
Metrics Filters Include: ["node_filesystem_*"]
Metrics Filters Exclude: ["node_filesystem_device_*"]
```

The configuration above will include only metrics that match `node_filesystem_*` pattern and do not match `node_filesystem_device_*`.

To keep only specific metrics, anchor the start and the end of the regexp of each metric:

> - the caret ^ matches the beginning of a text or line,
> - the dollar sign $ matches the end of a text.

Reviewer 3: ```suggestion
- caret anchor `^` matches the beginning of a text
- dollar anchor `$` matches the end of a text
```

---

If really want to improve this doc and make it more explicit, we can refer this: https://learn.microsoft.com/en-us/dotnet/standard/base-types/anchors-in-regular-expressions

For multiline mode, we have not written anything. But for regex anchors, it's understood. So should we make it more explicit or not? What do you think?

```yml
Metrics Filters Include: ["^node_network_net_dev_group$", "^node_network_up$"]
```
### Datastream Dataset Name

> The users of the Prometheus Input Package have the option of adding their own dataset name, to which the events get added. Prometheus Metrics from different services can be collected by adding multiple instances of Input package. The metrics can be filtered on the basis of dataset name.

Reviewer 3: ```suggestion

The users of the Prometheus Input Package have the option of adding their own dataset name, to which the events get added. Prometheus Metrics from different services can be collected by adding multiple instances of the Input package. The metrics can be filtered on the basis of the dataset name.
```
